fms/modules/tp.py

```diff
@@ -0,0 +1,84 @@
+import itertools
+from abc import ABCMeta, abstractmethod
+from typing import List, Type
+
+import torch
+import torch.nn as nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from fms.distributed.tensorparallel import (
+    apply_colwise_tp,
+    apply_embedding_tp,
+    apply_rowwise_tp,
+)
+
+
+class TPModule(nn.Module, metaclass=ABCMeta):
+    """
+    This is an abstract class that any nn.Module can implement to enable
+    Tensor Parallel. On top of inheriting from this class, the TP module
+    will have to implement list_colwise_weights, list_rowwise_weights,
+    list_embedding_weights, and import_module for their relevant weights.
+    Finally, the module must call setup_tp at the end of their __init__
+    function. See examples in attention.py, feedforward.py and embedding.py
+
+    """
+
+    rank: int
+    world_size: int
+
+    def setup_tp(self, rank: int, world_size: int) -> None:
+        self.rank = rank
+        self.world_size = world_size
+
+    def list_colwise_weights(self) -> List[str]:
+        return []
+
+    def list_rowwise_weights(self) -> List[str]:
+        return []
+
+    def list_embedding_weights(self) -> List[str]:
+        return []
+
+    @staticmethod
+    @abstractmethod
+    def import_module(module, group: ProcessGroup):
```

Comment:
i find this so weird since most languages don't support abstract static, but apparently python does (since 3.3).

Reply:
right... although I guess it is the best way to say "you need to implement a factory function with this interface for each subclass"